src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -111685,5 +111685,184 @@ void M(object? o1)
+        [InlineData("locationNonNull", "valueMaybeNull", "comparandMaybeNull", false)]
+        [InlineData("locationNonNull", "valueMaybeNull", "null", false)]
+        [InlineData("locationNonNull", "null", "comparandNonNull", false)]
+        [InlineData("locationNonNull", "null", "comparandMaybeNull", false)]
+        [InlineData("locationNonNull", "null", "null", false)]
+        [InlineData("locationMaybeNull", "valueNonNull", "comparandNonNull", false)]
+        [InlineData("locationMaybeNull", "valueNonNull", "comparandMaybeNull", false)]
+        [InlineData("locationMaybeNull", "valueNonNull", "null", true)]
+        [InlineData("locationMaybeNull", "valueMaybeNull", "comparandNonNull", false)]
+        [InlineData("locationMaybeNull", "valueMaybeNull", "comparandMaybeNull", false)]
+        [InlineData("locationMaybeNull", "valueMaybeNull", "null", false)]
+        [InlineData("locationMaybeNull", "null", "comparandNonNull", false)]
+        [InlineData("locationMaybeNull", "null", "comparandMaybeNull", false)]
+        [InlineData("locationMaybeNull", "null", "null", false)]
+        [WorkItem(36911, "https://github.com/dotnet/roslyn/issues/36911")]
+        public void CompareExchange_LocationNullState(string location, string value, string comparand, bool isLocationNonNull)
+        {
+            var source = $@"
+#pragma warning disable 0436
+using System.Threading;
+namespace System.Threading
+{{
+	public static class Interlocked
+    {{
+        public static object? CompareExchange(ref object? location, object? value, object? comparand) {{ return location; }}
+    }}
+}}
+
+class C
+{{
+    void M(object? locationNonNull, object? locationMaybeNull, object comparandNonNull, object? comparandMaybeNull, object valueNonNull, object? valueMaybeNull)
+    {{
+        locationNonNull = new object();
+        Interlocked.CompareExchange(ref {location}, {value}, {comparand});
+        _ = {location}.ToString();
+    }}
+}}
+    ";
+            var comp = CreateCompilation(source, options: WithNonNullTypesTrue());
+            comp.VerifyDiagnostics(isLocationNonNull
+                ? Array.Empty<DiagnosticDescription>()
+                : new[]
+                {
+                    // (18,13): warning CS8602: Dereference of a possibly null reference.
+                    //         _ = {location}.ToString();
+                    Diagnostic(ErrorCode.WRN_NullReferenceReceiver, location).WithLocation(18, 13)
+                });
+
+            source = $@"
+#pragma warning disable 0436
+using System.Threading;
+namespace System.Threading
+{{
+	public static class Interlocked
+    {{
+        public static T CompareExchange<T>(ref T location, T value, T comparand) where T : class? {{ return location; }}
+    }}
+}}
+
+class C
+{{
+    void M<T>(T? locationNonNull, T? locationMaybeNull, T comparandNonNull, T? comparandMaybeNull, T valueNonNull, T? valueMaybeNull) where T : class, new()
```

Comment:
>T? [](start = 14, length = 2)

Should this be not null?

Reply:
Changing to `T locationNonNull` introduces an additional warning when used in combination with any other arguments that are maybe-null. e.g.:

```cs
// (17,41): warning CS8600: Converting null literal or possible null value to non-nullable type.
//         Interlocked.CompareExchange(ref locationNonNull, valueNonNull, comparandMaybeNull);
Diagnostic(ErrorCode.WRN_ConvertingNullableToNonNullable, "locationNonNull").WithLocation(17, 41)
```

Reply:
I'd call that a good thing, not a bad thing.

---
In reply to: [301306204](https://github.com/dotnet/roslyn/pull/37014#discussion_r301306204) [](ancestors = 301306204)

Reply:
Added some tests which use `T location` because modifying the existing test to use an unannotated location  introduces excessive complexity.